Register `woocommerce-google-analytics-integration` script ASAP

tomalec opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Register `woocommerce-google-analytics-integration` script ASAP 
so other extensions (GLA) could add their data when needed.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/2297#issuecomment-2007805243


### Checks:
<!-- Mark completed items with an [x] -->
* [x] Does your code follow the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Screenshots:

<!--- Optional --->


### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Follow the steps from https://github.com/woocommerce/google-listings-and-ads/pull/2297
2. Or call 
```php
wp_add_inline_script(
    'woocommerce-google-analytics-integration',
	'alert("works");'
);
```
before `wp_enqueue_scripts`
4. Check that alert works.


### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

> Tweak - Register `woocommerce-google-analytics-integration` script earlier, so other extensions can plug in.

tomalec committed:
Register `woocommerce-google-analytics-integration` script ASAP

so other extensions (GLA) could add their data when needed.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/2297#issuecomment-2007805243

tomalec committed:
Add E2Es for GLA-like integration

eason9487 commented on tests/e2e/test-snippets/test-snippets.php:
```diff
@@ -23,3 +23,16 @@ function ( $modes ) {
 );
+
+/*
+ * Mimic the behavior of GLA, or other plugin adding some inline events before `wp_enqueue_scripts`.
```
Maybe using "Google Listings & Ads" instead of an abbreviation would be a bit better.

eason9487 reviewed (approved):
Thanks for the work. LGTM.

I tried to reproduce the issue but had no luck for several hours. After more investigation, it was finally found that this is mainly because block themes such as Twenty Twenty-Two change the execution sequence of WP actions against the traditional theme like Storefront.

The key reversal of sequence is that in a traditional theme, `wp_enqueue_scripts` is run before `woocommerce_after_single_product`, but it's reversed with a block theme.

I would suggest adding code comments about this finding to the code where relevant.

#### Execution sequence with Storefront:
- `wp_enqueue_scripts`: No. 258
- `woocommerce_before_single_product`: No. 301

<details>
<summary>
Full execution sequence
</summary>

1. woocommerce_customer_loaded
1. woocommerce_init
1. widgets_init
1. register_sidebar
1. register_sidebar
1. register_sidebar
1. register_sidebar
1. register_sidebar
1. register_sidebar
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. action_scheduler_init
1. woocommerce_register_taxonomy
1. registered_taxonomy
1. registered_taxonomy_product_type
1. registered_taxonomy
1. registered_taxonomy_product_visibility
1. registered_taxonomy
1. registered_taxonomy_product_cat
1. registered_taxonomy
1. registered_taxonomy_product_tag
1. registered_taxonomy
1. registered_taxonomy_product_shipping_class
1. registered_taxonomy
1. registered_taxonomy_pa_color
1. registered_taxonomy
1. registered_taxonomy_pa_size
1. woocommerce_after_register_taxonomy
1. woocommerce_register_post_type
1. registered_post_type
1. registered_post_type_product
1. registered_post_type
1. registered_post_type_product_variation
1. registered_post_type
1. registered_post_type_shop_order
1. registered_post_type
1. registered_post_type_shop_order_refund
1. registered_post_type
1. registered_post_type_shop_coupon
1. woocommerce_after_register_post_type
1. registered_post_type
1. registered_post_type_shop_order_placehold
1. woocommerce_blocks_payment_method_type_registration
1. wp_sitemaps_init
1. wp_default_styles
1. load_textdomain
1. load_textdomain
1. load_textdomain
1. registered_post_type
1. registered_post_type_patterns_ai_data
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. posts_selection
1. wp_cache_set_last_changed
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. posts_selection
1. woocommerce_blocks_active-filters_block_registration
1. woocommerce_blocks_add-to-cart-form_block_registration
1. woocommerce_blocks_all-products_block_registration
1. woocommerce_blocks_all-reviews_block_registration
1. woocommerce_blocks_attribute-filter_block_registration
1. woocommerce_blocks_breadcrumbs_block_registration
1. woocommerce_blocks_catalog-sorting_block_registration
1. woocommerce_blocks_legacy-template_block_registration
1. woocommerce_blocks_classic-shortcode_block_registration
1. woocommerce_blocks_customer-account_block_registration
1. woocommerce_blocks_featured-category_block_registration
1. woocommerce_blocks_featured-product_block_registration
1. woocommerce_blocks_filter-wrapper_block_registration
1. woocommerce_blocks_handpicked-products_block_registration
1. woocommerce_blocks_mini-cart_block_registration
1. woocommerce_blocks_store-notices_block_registration
1. woocommerce_blocks_price-filter_block_registration
1. woocommerce_blocks_product-add-to-cart_block_registration
1. woocommerce_blocks_product-best-sellers_block_registration
1. woocommerce_blocks_product-button_block_registration
1. woocommerce_blocks_product-categories_block_registration
1. woocommerce_blocks_product-category_block_registration
1. woocommerce_blocks_product-collection_block_registration
1. woocommerce_blocks_product-collection-no-results_block_registration
1. woocommerce_blocks_product-gallery_block_registration
1. woocommerce_blocks_product-gallery-large-image_block_registration
1. woocommerce_blocks_product-gallery-large-image-next-previous_block_registration
1. woocommerce_blocks_product-gallery-pager_block_registration
1. woocommerce_blocks_product-gallery-thumbnails_block_registration
1. woocommerce_blocks_product-image_block_registration
1. woocommerce_blocks_product-image-gallery_block_registration
1. woocommerce_blocks_product-new_block_registration
1. woocommerce_blocks_product-on-sale_block_registration
1. woocommerce_blocks_product-price_block_registration
1. woocommerce_blocks_product-template_block_registration
1. woocommerce_blocks_product-query_block_registration
1. woocommerce_blocks_product-average-rating_block_registration
1. woocommerce_blocks_product-rating_block_registration
1. woocommerce_blocks_product-rating-counter_block_registration
1. woocommerce_blocks_product-rating-stars_block_registration
1. woocommerce_blocks_product-results-count_block_registration
1. woocommerce_blocks_product-reviews_block_registration
1. woocommerce_blocks_product-sale-badge_block_registration
1. woocommerce_blocks_product-search_block_registration
1. woocommerce_blocks_product-sku_block_registration
1. woocommerce_blocks_product-stock-indicator_block_registration
1. woocommerce_blocks_product-summary_block_registration
1. woocommerce_blocks_product-tag_block_registration
1. woocommerce_blocks_product-title_block_registration
1. woocommerce_blocks_product-top-rated_block_registration
1. woocommerce_blocks_products-by-attribute_block_registration
1. woocommerce_blocks_rating-filter_block_registration
1. woocommerce_blocks_reviews-by-category_block_registration
1. woocommerce_blocks_reviews-by-product_block_registration
1. woocommerce_blocks_related-products_block_registration
1. woocommerce_blocks_product-details_block_registration
1. woocommerce_blocks_single-product_block_registration
1. woocommerce_blocks_stock-filter_block_registration
1. woocommerce_blocks_page-content-wrapper_block_registration
1. woocommerce_blocks_order-confirmation-status_block_registration
1. woocommerce_blocks_order-confirmation-summary_block_registration
1. woocommerce_blocks_order-confirmation-totals_block_registration
1. woocommerce_blocks_order-confirmation-totals-wrapper_block_registration
1. woocommerce_blocks_order-confirmation-downloads_block_registration
1. woocommerce_blocks_order-confirmation-downloads-wrapper_block_registration
1. woocommerce_blocks_order-confirmation-billing-address_block_registration
1. woocommerce_blocks_order-confirmation-shipping-address_block_registration
1. woocommerce_blocks_order-confirmation-billing-wrapper_block_registration
1. woocommerce_blocks_order-confirmation-shipping-wrapper_block_registration
1. woocommerce_blocks_order-confirmation-additional-information_block_registration
1. woocommerce_blocks_order-confirmation-additional-fields-wrapper_block_registration
1. woocommerce_blocks_order-confirmation-additional-fields_block_registration
1. woocommerce_blocks_cart_block_registration
1. woocommerce_blocks_cart-order-summary-taxes-block_block_registration
1. woocommerce_blocks_cart-order-summary-subtotal-block_block_registration
1. woocommerce_blocks_filled-cart-block_block_registration
1. woocommerce_blocks_empty-cart-block_block_registration
1. woocommerce_blocks_cart-totals-block_block_registration
1. woocommerce_blocks_cart-items-block_block_registration
1. woocommerce_blocks_cart-line-items-block_block_registration
1. woocommerce_blocks_cart-order-summary-block_block_registration
1. woocommerce_blocks_cart-express-payment-block_block_registration
1. woocommerce_blocks_proceed-to-checkout-block_block_registration
1. woocommerce_blocks_cart-accepted-payment-methods-block_block_registration
1. woocommerce_blocks_cart-order-summary-coupon-form-block_block_registration
1. woocommerce_blocks_cart-order-summary-discount-block_block_registration
1. woocommerce_blocks_cart-order-summary-fee-block_block_registration
1. woocommerce_blocks_cart-order-summary-heading-block_block_registration
1. woocommerce_blocks_cart-order-summary-shipping-block_block_registration
1. woocommerce_blocks_cart-cross-sells-block_block_registration
1. woocommerce_blocks_cart-cross-sells-products-block_block_registration
1. woocommerce_blocks_checkout_block_registration
1. woocommerce_blocks_checkout-actions-block_block_registration
1. woocommerce_blocks_checkout-additional-information-block_block_registration
1. woocommerce_blocks_checkout-billing-address-block_block_registration
1. woocommerce_blocks_checkout-contact-information-block_block_registration
1. woocommerce_blocks_checkout-express-payment-block_block_registration
1. woocommerce_blocks_checkout-fields-block_block_registration
1. woocommerce_blocks_checkout-order-note-block_block_registration
1. woocommerce_blocks_checkout-order-summary-block_block_registration
1. woocommerce_blocks_checkout-order-summary-cart-items-block_block_registration
1. woocommerce_blocks_checkout-order-summary-coupon-form-block_block_registration
1. woocommerce_blocks_checkout-order-summary-discount-block_block_registration
1. woocommerce_blocks_checkout-order-summary-fee-block_block_registration
1. woocommerce_blocks_checkout-order-summary-shipping-block_block_registration
1. woocommerce_blocks_checkout-order-summary-subtotal-block_block_registration
1. woocommerce_blocks_checkout-order-summary-taxes-block_block_registration
1. woocommerce_blocks_checkout-payment-block_block_registration
1. woocommerce_blocks_checkout-shipping-address-block_block_registration
1. woocommerce_blocks_checkout-shipping-methods-block_block_registration
1. woocommerce_blocks_checkout-shipping-method-block_block_registration
1. woocommerce_blocks_checkout-pickup-options-block_block_registration
1. woocommerce_blocks_checkout-terms-block_block_registration
1. woocommerce_blocks_checkout-totals-block_block_registration
1. woocommerce_blocks_mini-cart-contents_block_registration
1. woocommerce_blocks_empty-mini-cart-contents-block_block_registration
1. woocommerce_blocks_filled-mini-cart-contents-block_block_registration
1. woocommerce_blocks_mini-cart-footer-block_block_registration
1. woocommerce_blocks_mini-cart-items-block_block_registration
1. woocommerce_blocks_mini-cart-products-table-block_block_registration
1. woocommerce_blocks_mini-cart-shopping-button-block_block_registration
1. woocommerce_blocks_mini-cart-cart-button-block_block_registration
1. woocommerce_blocks_mini-cart-checkout-button-block_block_registration
1. woocommerce_blocks_mini-cart-title-block_block_registration
1. woocommerce_blocks_mini-cart-title-items-counter-block_block_registration
1. woocommerce_blocks_mini-cart-title-label-block_block_registration
1. wp_loaded
1. woocommerce_load_cart_from_session
1. parse_term_query
1. pre_get_terms
1. wp_cache_set_last_changed
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. woocommerce_cart_loaded_from_session
1. woocommerce_cart_updated
1. add_option
1. add_option__transient_doing_cron
1. added_option
1. set_transient_doing_cron
1. setted_transient
1. requests-requests.before_request
1. requests-curl.before_request
1. http_api_curl
1. requests-curl.before_send
1. requests-curl.after_send
1. requests-curl.after_request
1. requests-requests.before_parse
1. http_api_debug
1. parse_request
1. parse_query
1. pre_get_posts
1. posts_selection
1. metadata_lazyloader_queued_objects
1. send_headers
1. wp
1. woocommerce_set_cart_cookies
1. template_redirect
1. admin_bar_init
1. add_admin_bar_menus
1. get_header
1. wp_before_load_template
1. wp_head
1. **wp_enqueue_scripts** 📌📌📌
1. enqueue_block_assets
1. wp_print_styles
1. wp_print_scripts
1. parse_term_query
1. pre_get_terms
1. wp_body_open
1. admin_bar_menu
1. wp_before_admin_bar_render
1. wp_after_admin_bar_render
1. storefront_before_site
1. storefront_before_header
1. storefront_header
1. the_widget
1. pre_get_product_search_form
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. posts_selection
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. posts_selection
1. the_widget
1. storefront_before_content
1. parse_term_query
1. pre_get_terms
1. woocommerce_breadcrumb
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. storefront_content_top
1. wp_after_load_template
1. woocommerce_before_main_content
1. loop_start
1. the_post
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. wp_before_load_template
1. **woocommerce_before_single_product** 📌📌📌
1. woocommerce_before_single_product_summary
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. woocommerce_product_thumbnails
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_after_template_part
1. woocommerce_single_product_summary
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_simple_add_to_cart
1. woocommerce_before_template_part
1. woocommerce_before_add_to_cart_form
1. woocommerce_before_add_to_cart_button
1. woocommerce_before_add_to_cart_quantity
1. woocommerce_before_template_part
1. woocommerce_before_quantity_input_field
1. woocommerce_after_quantity_input_field
1. woocommerce_after_template_part
1. woocommerce_after_add_to_cart_quantity
1. woocommerce_after_add_to_cart_button
1. woocommerce_after_add_to_cart_form
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. woocommerce_product_meta_start
1. metadata_lazyloader_queued_objects
1. woocommerce_product_meta_end
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. woocommerce_share
1. woocommerce_after_template_part
1. woocommerce_after_single_product_summary
1. woocommerce_before_template_part
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. woocommerce_product_additional_information
1. parse_term_query
1. pre_get_terms
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_after_template_part
1. parse_comment_query
1. pre_get_comments
1. wp_cache_set_last_changed
1. comment_form_before
1. comment_form_top
1. comment_form_logged_in_after
1. comment_form
1. comment_form_after
1. woocommerce_product_after_tabs
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. the_post
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. woocommerce_after_template_part
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. woocommerce_product_read
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. woocommerce_product_read
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. woocommerce_product_read
1. woocommerce_before_template_part
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. the_post
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. wp_before_load_template
1. woocommerce_before_shop_loop_item
1. woocommerce_before_shop_loop_item_title
1. woocommerce_shop_loop_item_title
1. woocommerce_after_shop_loop_item_title
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_after_shop_loop_item
1. woocommerce_before_template_part
1. parse_term_query
1. pre_get_terms
1. woocommerce_after_template_part
1. wp_after_load_template
1. the_post
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. wp_before_load_template
1. woocommerce_before_shop_loop_item
1. woocommerce_before_shop_loop_item_title
1. woocommerce_shop_loop_item_title
1. woocommerce_after_shop_loop_item_title
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_after_shop_loop_item
1. woocommerce_before_template_part
1. parse_term_query
1. pre_get_terms
1. woocommerce_after_template_part
1. wp_after_load_template
1. the_post
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. wp_before_load_template
1. woocommerce_before_shop_loop_item
1. woocommerce_before_shop_loop_item_title
1. woocommerce_shop_loop_item_title
1. woocommerce_after_shop_loop_item_title
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_after_shop_loop_item
1. woocommerce_before_template_part
1. parse_term_query
1. pre_get_terms
1. woocommerce_after_template_part
1. wp_after_load_template
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. the_post
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. woocommerce_after_template_part
1. parse_term_query
1. pre_get_terms
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. parse_term_query
1. pre_get_terms
1. metadata_lazyloader_queued_objects
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. parse_term_query
1. pre_get_terms
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. woocommerce_after_single_product
1. wp_after_load_template
1. loop_end
1. woocommerce_after_main_content
1. storefront_sidebar
1. get_sidebar
1. wp_before_load_template
1. dynamic_sidebar_before
1. dynamic_sidebar
1. dynamic_sidebar
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. posts_selection
1. parse_term_query
1. pre_get_terms
1. metadata_lazyloader_queued_objects
1. dynamic_sidebar
1. parse_comment_query
1. pre_get_comments
1. metadata_lazyloader_queued_objects
1. dynamic_sidebar
1. dynamic_sidebar
1. parse_term_query
1. pre_get_terms
1. metadata_lazyloader_queued_objects
1. dynamic_sidebar_after
1. wp_after_load_template
1. woocommerce_sidebar
1. get_footer
1. wp_before_load_template
1. storefront_before_footer
1. storefront_footer
1. the_widget
1. pre_get_product_search_form
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. storefront_after_footer
1. wp_footer
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. wp_print_footer_scripts
1. wp_after_load_template
1. shutdown

</details>

#### Execution sequence  with Twenty Twenty-Two:

- `woocommerce_before_single_product`: No. 605
- `wp_enqueue_scripts`: No. 621

<details>
<summary>Full execution sequence</summary>

1. woocommerce_customer_loaded
1. woocommerce_init
1. widgets_init
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. wp_register_sidebar_widget
1. action_scheduler_init
1. woocommerce_register_taxonomy
1. registered_taxonomy
1. registered_taxonomy_product_type
1. registered_taxonomy
1. registered_taxonomy_product_visibility
1. registered_taxonomy
1. registered_taxonomy_product_cat
1. registered_taxonomy
1. registered_taxonomy_product_tag
1. registered_taxonomy
1. registered_taxonomy_product_shipping_class
1. registered_taxonomy
1. registered_taxonomy_pa_color
1. registered_taxonomy
1. registered_taxonomy_pa_size
1. woocommerce_after_register_taxonomy
1. woocommerce_register_post_type
1. registered_post_type
1. registered_post_type_product
1. registered_post_type
1. registered_post_type_product_variation
1. registered_post_type
1. registered_post_type_shop_order
1. registered_post_type
1. registered_post_type_shop_order_refund
1. registered_post_type
1. registered_post_type_shop_coupon
1. woocommerce_after_register_post_type
1. registered_post_type
1. registered_post_type_shop_order_placehold
1. woocommerce_blocks_payment_method_type_registration
1. wp_default_styles
1. wp_sitemaps_init
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. wp_cache_set_last_changed
1. posts_selection
1. wp_cache_set_last_changed
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. load_textdomain
1. load_textdomain
1. load_textdomain
1. registered_post_type
1. registered_post_type_patterns_ai_data
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. posts_selection
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. posts_selection
1. woocommerce_blocks_active-filters_block_registration
1. woocommerce_blocks_add-to-cart-form_block_registration
1. woocommerce_blocks_all-products_block_registration
1. woocommerce_blocks_all-reviews_block_registration
1. woocommerce_blocks_attribute-filter_block_registration
1. woocommerce_blocks_breadcrumbs_block_registration
1. woocommerce_blocks_catalog-sorting_block_registration
1. woocommerce_blocks_legacy-template_block_registration
1. woocommerce_blocks_classic-shortcode_block_registration
1. woocommerce_blocks_customer-account_block_registration
1. woocommerce_blocks_featured-category_block_registration
1. woocommerce_blocks_featured-product_block_registration
1. woocommerce_blocks_filter-wrapper_block_registration
1. woocommerce_blocks_handpicked-products_block_registration
1. woocommerce_blocks_mini-cart_block_registration
1. woocommerce_blocks_store-notices_block_registration
1. woocommerce_blocks_price-filter_block_registration
1. woocommerce_blocks_product-add-to-cart_block_registration
1. woocommerce_blocks_product-best-sellers_block_registration
1. woocommerce_blocks_product-button_block_registration
1. woocommerce_blocks_product-categories_block_registration
1. woocommerce_blocks_product-category_block_registration
1. woocommerce_blocks_product-collection_block_registration
1. woocommerce_blocks_product-collection-no-results_block_registration
1. woocommerce_blocks_product-gallery_block_registration
1. woocommerce_blocks_product-gallery-large-image_block_registration
1. woocommerce_blocks_product-gallery-large-image-next-previous_block_registration
1. woocommerce_blocks_product-gallery-pager_block_registration
1. woocommerce_blocks_product-gallery-thumbnails_block_registration
1. woocommerce_blocks_product-image_block_registration
1. woocommerce_blocks_product-image-gallery_block_registration
1. woocommerce_blocks_product-new_block_registration
1. woocommerce_blocks_product-on-sale_block_registration
1. woocommerce_blocks_product-price_block_registration
1. woocommerce_blocks_product-template_block_registration
1. woocommerce_blocks_product-query_block_registration
1. woocommerce_blocks_product-average-rating_block_registration
1. woocommerce_blocks_product-rating_block_registration
1. woocommerce_blocks_product-rating-counter_block_registration
1. woocommerce_blocks_product-rating-stars_block_registration
1. woocommerce_blocks_product-results-count_block_registration
1. woocommerce_blocks_product-reviews_block_registration
1. woocommerce_blocks_product-sale-badge_block_registration
1. woocommerce_blocks_product-search_block_registration
1. woocommerce_blocks_product-sku_block_registration
1. woocommerce_blocks_product-stock-indicator_block_registration
1. woocommerce_blocks_product-summary_block_registration
1. woocommerce_blocks_product-tag_block_registration
1. woocommerce_blocks_product-title_block_registration
1. woocommerce_blocks_product-top-rated_block_registration
1. woocommerce_blocks_products-by-attribute_block_registration
1. woocommerce_blocks_rating-filter_block_registration
1. woocommerce_blocks_reviews-by-category_block_registration
1. woocommerce_blocks_reviews-by-product_block_registration
1. woocommerce_blocks_related-products_block_registration
1. woocommerce_blocks_product-details_block_registration
1. woocommerce_blocks_single-product_block_registration
1. woocommerce_blocks_stock-filter_block_registration
1. woocommerce_blocks_page-content-wrapper_block_registration
1. woocommerce_blocks_order-confirmation-status_block_registration
1. woocommerce_blocks_order-confirmation-summary_block_registration
1. woocommerce_blocks_order-confirmation-totals_block_registration
1. woocommerce_blocks_order-confirmation-totals-wrapper_block_registration
1. woocommerce_blocks_order-confirmation-downloads_block_registration
1. woocommerce_blocks_order-confirmation-downloads-wrapper_block_registration
1. woocommerce_blocks_order-confirmation-billing-address_block_registration
1. woocommerce_blocks_order-confirmation-shipping-address_block_registration
1. woocommerce_blocks_order-confirmation-billing-wrapper_block_registration
1. woocommerce_blocks_order-confirmation-shipping-wrapper_block_registration
1. woocommerce_blocks_order-confirmation-additional-information_block_registration
1. woocommerce_blocks_order-confirmation-additional-fields-wrapper_block_registration
1. woocommerce_blocks_order-confirmation-additional-fields_block_registration
1. woocommerce_blocks_cart_block_registration
1. woocommerce_blocks_cart-order-summary-taxes-block_block_registration
1. woocommerce_blocks_cart-order-summary-subtotal-block_block_registration
1. woocommerce_blocks_filled-cart-block_block_registration
1. woocommerce_blocks_empty-cart-block_block_registration
1. woocommerce_blocks_cart-totals-block_block_registration
1. woocommerce_blocks_cart-items-block_block_registration
1. woocommerce_blocks_cart-line-items-block_block_registration
1. woocommerce_blocks_cart-order-summary-block_block_registration
1. woocommerce_blocks_cart-express-payment-block_block_registration
1. woocommerce_blocks_proceed-to-checkout-block_block_registration
1. woocommerce_blocks_cart-accepted-payment-methods-block_block_registration
1. woocommerce_blocks_cart-order-summary-coupon-form-block_block_registration
1. woocommerce_blocks_cart-order-summary-discount-block_block_registration
1. woocommerce_blocks_cart-order-summary-fee-block_block_registration
1. woocommerce_blocks_cart-order-summary-heading-block_block_registration
1. woocommerce_blocks_cart-order-summary-shipping-block_block_registration
1. woocommerce_blocks_cart-cross-sells-block_block_registration
1. woocommerce_blocks_cart-cross-sells-products-block_block_registration
1. woocommerce_blocks_checkout_block_registration
1. woocommerce_blocks_checkout-actions-block_block_registration
1. woocommerce_blocks_checkout-additional-information-block_block_registration
1. woocommerce_blocks_checkout-billing-address-block_block_registration
1. woocommerce_blocks_checkout-contact-information-block_block_registration
1. woocommerce_blocks_checkout-express-payment-block_block_registration
1. woocommerce_blocks_checkout-fields-block_block_registration
1. woocommerce_blocks_checkout-order-note-block_block_registration
1. woocommerce_blocks_checkout-order-summary-block_block_registration
1. woocommerce_blocks_checkout-order-summary-cart-items-block_block_registration
1. woocommerce_blocks_checkout-order-summary-coupon-form-block_block_registration
1. woocommerce_blocks_checkout-order-summary-discount-block_block_registration
1. woocommerce_blocks_checkout-order-summary-fee-block_block_registration
1. woocommerce_blocks_checkout-order-summary-shipping-block_block_registration
1. woocommerce_blocks_checkout-order-summary-subtotal-block_block_registration
1. woocommerce_blocks_checkout-order-summary-taxes-block_block_registration
1. woocommerce_blocks_checkout-payment-block_block_registration
1. woocommerce_blocks_checkout-shipping-address-block_block_registration
1. woocommerce_blocks_checkout-shipping-methods-block_block_registration
1. woocommerce_blocks_checkout-shipping-method-block_block_registration
1. woocommerce_blocks_checkout-pickup-options-block_block_registration
1. woocommerce_blocks_checkout-terms-block_block_registration
1. woocommerce_blocks_checkout-totals-block_block_registration
1. woocommerce_blocks_mini-cart-contents_block_registration
1. woocommerce_blocks_empty-mini-cart-contents-block_block_registration
1. woocommerce_blocks_filled-mini-cart-contents-block_block_registration
1. woocommerce_blocks_mini-cart-footer-block_block_registration
1. woocommerce_blocks_mini-cart-items-block_block_registration
1. woocommerce_blocks_mini-cart-products-table-block_block_registration
1. woocommerce_blocks_mini-cart-shopping-button-block_block_registration
1. woocommerce_blocks_mini-cart-cart-button-block_block_registration
1. woocommerce_blocks_mini-cart-checkout-button-block_block_registration
1. woocommerce_blocks_mini-cart-title-block_block_registration
1. woocommerce_blocks_mini-cart-title-items-counter-block_block_registration
1. woocommerce_blocks_mini-cart-title-label-block_block_registration
1. wp_loaded
1. woocommerce_load_cart_from_session
1. parse_term_query
1. pre_get_terms
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. woocommerce_cart_loaded_from_session
1. woocommerce_cart_updated
1. parse_request
1. parse_query
1. pre_get_posts
1. posts_selection
1. metadata_lazyloader_queued_objects
1. send_headers
1. wp
1. woocommerce_set_cart_cookies
1. template_redirect
1. admin_bar_init
1. add_admin_bar_menus
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. loop_no_results
1. render_block_core_template_part_file
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. posts_selection
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. posts_selection
1. rest_api_init
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. register_setting
1. woocommerce_email
1. wc_payment_gateways_initialized
1. register_setting
1. register_setting
1. woocommerce_blocks_cart_enqueue_data
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. parse_term_query
1. pre_get_terms
1. woocommerce_breadcrumb
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. woocommerce_product_thumbnails
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_after_template_part
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. woocommerce_before_single_product_summary
1. woocommerce_single_product_summary
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. woocommerce_simple_add_to_cart
1. woocommerce_before_template_part
1. woocommerce_before_add_to_cart_form
1. woocommerce_before_add_to_cart_button
1. woocommerce_before_add_to_cart_quantity
1. woocommerce_before_template_part
1. woocommerce_before_quantity_input_field
1. woocommerce_after_quantity_input_field
1. woocommerce_after_template_part
1. woocommerce_after_add_to_cart_quantity
1. woocommerce_after_add_to_cart_button
1. woocommerce_after_add_to_cart_form
1. woocommerce_after_template_part
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_meta_start
1. woocommerce_product_meta_end
1. loop_start
1. the_post
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. woocommerce_before_template_part
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_before_template_part
1. woocommerce_product_additional_information
1. parse_term_query
1. pre_get_terms
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. woocommerce_after_template_part
1. parse_comment_query
1. pre_get_comments
1. wp_cache_set_last_changed
1. comment_form_before
1. comment_form_top
1. comment_form_logged_in_after
1. comment_form
1. comment_form_after
1. woocommerce_product_after_tabs
1. woocommerce_after_template_part
1. loop_end
1. woocommerce_share
1. woocommerce_after_single_product_summary
1. woocommerce_before_template_part
1. the_post
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. woocommerce_after_template_part
1. parse_term_query
1. pre_get_terms
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. parse_term_query
1. pre_get_terms
1. metadata_lazyloader_queued_objects
1. update_option
1. update_option__transient_timeout_wc_related_35
1. updated_option
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. woocommerce_product_read
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. woocommerce_product_read
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. woocommerce_product_read
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. parse_term_query
1. pre_get_terms
1. woocommerce_product_read
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. posts_selection
1. parse_term_query
1. pre_get_terms
1. metadata_lazyloader_queued_objects
1. loop_start
1. the_post
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. parse_term_query
1. pre_get_terms
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. the_post
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. parse_term_query
1. pre_get_terms
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. the_post
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. parse_term_query
1. pre_get_terms
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. the_post
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. parse_term_query
1. pre_get_terms
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. loop_end
1. the_post
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. parse_term_query
1. pre_get_terms
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. metadata_lazyloader_queued_objects
1. metadata_lazyloader_queued_objects
1. woocommerce_product_read
1. woocommerce_before_main_content
1. **woocommerce_before_single_product** 📌📌📌
1. parse_term_query
1. pre_get_terms
1. woocommerce_after_single_product
1. woocommerce_after_main_content
1. woocommerce_sidebar
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. loop_no_results
1. render_block_core_template_part_file
1. wp_head
1. **wp_enqueue_scripts** 📌📌📌
1. enqueue_block_assets
1. wp_print_styles
1. wp_print_scripts
1. wp_body_open
1. admin_bar_menu
1. wp_before_admin_bar_render
1. wp_after_admin_bar_render
1. wp_footer
1. woocommerce_before_template_part
1. woocommerce_after_template_part
1. wp_print_footer_scripts
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. woocommerce_shipping_init
1. parse_term_query
1. pre_get_terms
1. woocommerce_load_shipping_methods
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. parse_tax_query
1. parse_query
1. pre_get_posts
1. parse_tax_query
1. parse_term_query
1. pre_get_terms
1. posts_selection
1. shutdown

</details>

tomalec committed:
Merge branch 'trunk' into fix/register-timing

tomalec committed:
Resolve GLA abbrevation in the code comment

Address https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/404#discussion_r1540428835

tomalec committed:
Test inlining scripts with a wp_head,

to make sure it's run before `enqueue_scripts` regardless of the theme.

Address https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/404#pullrequestreview-1962236542

tomalec commented:
Thank you, @eason9487, for the detailed review and investigation!

> The key reversal of sequence is that in a traditional theme, wp_enqueue_scripts is run before woocommerce_after_single_product, but it's reversed with a block theme.

WDYT of adding the test snippet in a different hook - `wp_head` to make sure it's before `enqueue_scripts`? https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/404/commits/a24f14a9f19e20d6fcd0b9e78bf40fd3e7ad2e37
It's further from testing what we actually need in the real case: Make sure hooking for the `woocommerce_before_thankyou, woocommerce_after_single_product, wp_body_open` events works https://github.com/woocommerce/google-listings-and-ads/blob/88782dc7c89b1406d5ab30202921ca4ebb08f7b9/src/Google/GlobalSiteTag.php#L121-L140, but  less theme-fragile.

tomalec committed:
Fix whitespace for PHPCS

eason9487 commented on tests/e2e/test-snippets/test-snippets.php:
```diff
@@ -23,3 +23,20 @@ function ( $modes ) {
+ * adding some inline events before `wp_enqueue_scripts.`
+ */
+add_action(
+	'wp_head',
```
I tested with the following changes to simulate the implementation before this PR, but the test can pass (regardless of the theme types), which means the test case cannot assert the purpose.

```diff
diff --git a/includes/class-wc-google-gtag-js.php b/includes/class-wc-google-gtag-js.php
index 212e381..b3ee9e4 100644
--- a/includes/class-wc-google-gtag-js.php
+++ b/includes/class-wc-google-gtag-js.php
@@ -48,7 +48,7 @@ class WC_Google_Gtag_JS extends WC_Abstract_Google_Analytics_JS {
 
                $this->map_hooks();
 
-               $this->register_scripts();
+               // $this->register_scripts();
                // Setup frontend scripts
                add_action( 'wp_enqueue_scripts', array( $this, 'enquque_tracker' ), 5 );
                add_action( 'wp_footer', array( $this, 'inline_script_data' ) );
@@ -89,6 +89,7 @@ class WC_Google_Gtag_JS extends WC_Abstract_Google_Analytics_JS {
         * @return void
         */
        public function enquque_tracker(): void {
+               $this->register_scripts();
                wp_enqueue_script( 'google-tag-manager' );
                // tracker.js needs to be executed ASAP, the remaining bits for main.js could be deffered,
                // but to reduce the traffic, we ship it all together.
```

I guess it's because `wp_enqueue_scripts` is triggered by `wp_head`, so even `wp_head` is run first, callbacks registered to `wp_head` still be run after the ones registered to `wp_enqueue_scripts`.

![1](https://github.com/woocommerce/woocommerce-google-analytics-integration/assets/17420811/e06e81a1-af3d-42a6-86ff-8665d081c885)

I'm not sure which one is better, but it seems that the `wp` action at least ensures that its callbacks are always run earlier than `wp_enqueue_scripts`.

eason9487 reviewed (commented):
Thanks for the update. It might need a bit of adjustment to ensure the new test case can fulfill its purpose.

tomalec committed:
Use `wp` hook for testing early registration

to make sure it's way before `enqueue_scripts`

Address https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/404#pullrequestreview-1968488103

tomalec commented:
Great catch, @eason9487. Thanks!
